### PR TITLE
Optimizer graphbreak only if needed

### LINKED ISSUE #104053: Tracking issue for optimizer graph not being an inference graph
### 🐛 Describe the bug

The optimizer is incorrectly classified as a training graph in aot. The main reason this occurs is that some parameters have `requires_grad=True` *and* torchdynamo calls set_grad_enabled(True|False) while simultaneously putting a call_function node in the graph which performs the global state change. 

As a result, when dynamo traces a no_grad context manager, it will trace the *exit* of the context manager *before* calling the user compiler. As a result, by the time we reach aot, grad is enabled, and the graph is treated as a training graph.

Why this matters:
Functionalization appends an epilogue which aggregates all mutation updates of tensors into a single code block consisting of a bunch of copies. In inference mode, these copies are visible to inductor while in training mode, these copies are not visible to inductor. If the copies are not visible to inductor, inductor will allocate tensors for every single parameter in the graph in order for them to get sent to the epilogue, which is atrocious for perf. 

Solutions:
1. graph break before grad is re-enabled
2. in functionalization make the epilogue visible in training mode
3. with compiled backwards, the entire graph is an inference graph by default

1 is two lines of code, so that approach is the easiest. 
3 is getting implemented currently, so once it's ready we can remove the graph breaks.

### Error logs

N/A

### Minified repro

_No response_

### Versions

N/a

cc @ezyang @msaroufim @wconstab @bdhirsh @anijain2305 @zou3519

### PR DESCRIPTION
**Pull request by c-rizz**
Partially Fixes #104053

The graph breaks in [_use_grad_for_differentiable](https://github.com/pytorch/pytorch/blob/0ba68403ffcc5971be991113c47dccf490110cd1/torch/optim/optimizer.py#L80) introduce some unavoidable slowdowns when using torch.compile, but they are only necessary when the grad state needs to be changed. If the grad state required by the optimizer's "differentiable" flag is the same as the current one, which is common, the graph breaks are unnecessary. This PR adds a check for this case.

I am not sure if it is actually the case to try removing the graph breaks altogether. @mlazos 
See the discussion [here at issue 104053](https://github.com/pytorch/pytorch/issues/104053#issuecomment-3785672407)

### CODE REVIEW

File: torch/optim/optimizer.py [hunk 1]
```diff
@@ -63,6 +63,10 @@ def _use_grad(*args: _P.args, **kwargs: _P.kwargs) -> _T:
         # pyrefly: ignore [unsupported-operation]
         self = cast(Optimizer, args[0])  # assume first positional arg is `self`
         prev_grad = torch.is_grad_enabled()
+        needed_grad = self.defaults["differentiable"]
+        if needed_grad == prev_grad:
+            return func(*args, **kwargs)
```
Comment (janeyx99) [inline torch/optim/optimizer.py]: @mlazos changing grad mode will auto graph break right? So I believe these lines aren't necessary.

Comment (mlazos) [reply torch/optim/optimizer.py]: we stick these in the graph actually so this should work.
